### PR TITLE
schannel: offer a "best effort" revocation check mode

### PR DESCRIPTION
**Pull request by dscho**
When running e.g. with Fiddler, the schannel backend fails with an unhelpful error message:

```
Unknown error (0x80092012) - The revocation function was unable to check revocation for the certificate.
```

Unfortunately, it affects not only debugging tools like Fiddler, but a large number of enterprise setups where MITM proxies are the rule rather than the exception. So it's not like this is a niche problem.

This has been discussed in plenty of issues, e.g. in https://github.com/curl/curl/issues/3727 or in https://github.com/curl/curl/issues/264.

It has also been discussed plenty of times in the context of Git for Windows and of GitHub Desktop, as many of our users ran straight into these issues with proxies and/or certificates lacking distribution points for CRLs.

Hopefully we will get this (as far as cURL is concerned, still opt-in) behavior, at long last?

/cc @georgeok @jay @niik

@jay, I basically copy-edited the test case straight from https://github.com/georgeok/curl/commit/bd695516eecf3047aa8f1613fd2c8b644ef6463e and am fairly certain that this is not what we want. If I am reading it correctly, this test requires an internet connection and a not-further-specified proxy behind a public IP. I would _love_ to fix this, but I don't know how. Essentially, what I want is a test that verifies that schannel works even when using a proxy with a certificate that is missing a CRL distribution point. Would you kindly help me out here?

### CODE REVIEW

**Comment by bagder** [comment]
(I'll just note that this is too late for getting into 7.69.0)

**Comment by dscho** [comment]
Okay :-) I'll update the version numbers once we have a working regression test...

**Comment by jay** [comment]
We have a solution to this already, --ssl-no-revoke. I'm against soft revocation, I think it's misleading. That said I suggest calling this --ssl-soft-revoke instead of best effort. Best effort could be interpreted as the best way to revoke. "soft" is clearly less. Or split it up --ssl-ignore-revoke-missing --ssl-ignore-revoke-offline. Or how about George did something like that in #4078... would --ssl-ignore-revoke-missing not be enough? (Note a revoke missing error technically means revocation could not be performed, not necessarily because there is no revocation point.)

**Comment by dscho** [comment]
> We have a solution to this already, --ssl-no-revoke.

Unfortunately, this is not a solution at all...

I would not respond so forcefully to your claim if I had not had to field my share of support requests.

With `--ssl-no-revoke` you simply open a wide gaping hole. The suggested `--revoke-best-effort` tries to at least figure out whether the certificate has been revoked. If there was no URL given (in which case you really cannot do much of anything, can you?) or the server is offline (in which case you don't really know why, but should you then assume that everything is revoked? Probably not. Web browsers don't do that. OpenSSL-backed cURL does not do that. Only the Secure Channel backend does, contrary to your claimed goal to have as little backend-specific behavior as possible) there is not a whole lot you can ask the code to do. But if the server is _not_ offline, `-ssl-no-revoke` still ignores it.

This is not really acceptable for a vast majority of our users.

And if I had to field my share of support requests, the GitHub Desktop team had to field a dozen time as many.

To make things utterly clear: the situation is _so bad_ with the unfortunate lack of a best effort strategy in cURL that, should you dig in and defend your stance despite all of the user complaints to the contrary, you will force me to ship a modified cURL in Git for Windows.

I don't want to fork from cURL. I really don't. I hope you don't make me.

> I'm against soft revocation, I think it's misleading.

It has nothing to do with soft, please do not misrepresent what this does. It is a best effort. Don't take my word for it, take the word of another Microsoftie (if you don't trust me): https://github.com/curl/curl/issues/264#issuecomment-114087899.

In other words, this is not a soft revocation (what would that be, anyway? "soft" would suggest that you say "well, this is revoked, but this one time we let it pass anyway"?). This is a best effort. And it is the common practice of web browsers out there. Above-quoted Eric Lawrence is a Microsoft Edge developer. You might want to assume that he really knows what he is talking about.

> Or split it up --ssl-ignore-revoke-missing --ssl-ignore-revoke-offline.

I can do that. It will make it a bit harder to follow the lead of web browsers to implement a best effort strategy, but if that's what it takes to get this use case supported, so be it.

More importantly, could you guide me how to implement a test case for this?

**Comment by bagder** [comment]
> slit it up --ssl-ignore-revoke-missing --ssl-ignore-revoke-offline.

I don't understand why doing it as two separate options would make anything better.

I think @dscho makes a compelling argument for this feature and I don't see how adding two options for this is what anyone wants nor how it helps anyone - but instead just makes it more complicated. Revocations is a difficult and unfortunately mostly broken area of TLS and certificates. I don't think we do our users any benefits by refusing this PR.

**Comment by dscho** [comment]
@niik pointed out that I forgot the case where the certificates specified via `--cainfo` lack CRL distribution points, as is the case in the tests.

Speaking of tests, @niik also came up with a better test (i.e. one that does not try to access a proxy with a public IP) by copy-editing 310 to our need. I verified locally (with a newly-created ,MSYS2-packaged `tstunnel.exe`) that this works indeed as desired.

**Comment by dscho** [comment]
> (I'll just note that this is too late for getting into 7.69.0)

I changed the patch tentatively to refer to 7.70.0 instead.

Of course, I understand that 7.69.0 is due March 4th, so I hope that this PR will be rebase-merged soon thereafter :-)

**Comment by jay** [comment]
This is soft revocation. The revocation check for whatever reason can't be performed and so it fails. You want to soft-fail it. Hence soft revocation. You are making assumptions about what has caused the two errors you want to ignore. If an attacker can make the client ignore the revocation check it is in my _opinion_ not worth checking. I have discussed this already in the issues. I feel I've treated everyone fairly on this.

**Comment by dscho** [comment]
I feel reminded of the "WinSSL" vs "Secure Channel" issue: insisting on calling it something different than literally everybody else is just confusing.

I'd rather refrain from confusing users.

Same reason I'd like this to be *one* cURL flag instead of two, with a name that talks about the effect, rather than using names that *must* be less than obvious to occasional users.

File: lib/doh.c [hunk 1]
```diff
@@ -351,6 +354,8 @@ static CURLcode dohprobe(struct Curl_easy *data,
     }
     if(data->set.ssl.no_revoke)
       ERROR_CHECK_SETOPT(CURLOPT_SSL_OPTIONS, CURLSSLOPT_NO_REVOKE);
+    if(data->set.ssl.revoke_best_effort)
+      ERROR_CHECK_SETOPT(CURLOPT_SSL_OPTIONS, CURLSSLOPT_REVOKE_BEST_EFFORT);
```
Comment (jay) [inline lib/doh.c]: This and the proxy one in doh will need to be changed to `else if` because otherwise best effort overrides no revoke, think of it like this
curl_easy_setopt(curl, CURLOPT_SSL_OPTIONS, CURLSSLOPT_NO_REVOKE);
curl_easy_setopt(curl, CURLOPT_SSL_OPTIONS, CURLSSLOPT_REVOKE_BEST_EFFORT);

Comment (dscho) [reply lib/doh.c]: Yes, of course!

**Comment by dscho** [comment]
Hrm. I tried to detect Msys so that we can avoid testing 2070 there (it would always fail with "curl: (4) schannel: CA cert support not built in", there's nothing we can do about it), but apparently [this precheck](https://github.com/curl/curl/pull/4981/files#diff-827bc4e291e65dc617914f550696c8beR45) does not catch it:

```sh
curl -V | grep -q pc-mingw32 && echo "MINGW has no CA cert support"
```

Any ideas?

Also, the FreeBSD CI is failing due to some sort of relatively obvious version check mismatch. What is _not_ obvious to me is what I can do about it.

Lastly, there seem to be a ton of CI failures on Windows that however appear not to be caused by _this_ PR. Any guidance?

**Comment by mback2k** [comment]
> Also, the FreeBSD CI is failing due to some sort of relatively obvious version check mismatch. What is _not_ obvious to me is what I can do about it.

This should be fixed via 691b71be930f0e285c8f7a76efd56bbe0576cda6 in master now.

> Lastly, there seem to be a ton of CI failures on Windows that however appear not to be caused by _this_ PR. Any guidance?

These should be fixed via 9aaca09044de4d4116822f25d2cf9c780d7465ce. Please rebase. Also AppVeyor may still randomly fail some tests for currently unknown reason, then just retry by re-pushing or asking one of us to re-run the CI.

**Comment by mback2k** [comment]
> Hrm. I tried to detect Msys so that we can avoid testing 2070 there (it would always fail with "curl: (4) schannel: CA cert support not built in", there's nothing we can do about it), but apparently [this precheck](https://github.com/curl/curl/pull/4981/files#diff-827bc4e291e65dc617914f550696c8beR45) does not catch it:
> 
> ```shell
> curl -V | grep -q pc-mingw32 && echo "MINGW has no CA cert support"
> ```

Maybe do the check like the testsuite itself does?
https://github.com/curl/curl/blob/3c1b9145c78de3370c548ead90537c91d9803073/tests/runtests.pl#L2721

**Comment by dscho** [comment]
> > Hrm. I tried to detect Msys so that we can avoid testing 2070 there (it would always fail with "curl: (4) schannel: CA cert support not built in", there's nothing we can do about it), but apparently [this precheck](https://github.com/curl/curl/pull/4981/files#diff-827bc4e291e65dc617914f550696c8beR45) does not catch it:
> > ```shell
> > curl -V | grep -q pc-mingw32 && echo "MINGW has no CA cert support"
> > ```
> 
> Maybe do the check like the testsuite itself does?
> 
> https://github.com/curl/curl/blob/3c1b9145c78de3370c548ead90537c91d9803073/tests/runtests.pl#L2721

Ah! You pointed me in the right direction. I will add a new "feature" for the MinGW build and use that to exclude testing it in 2070.

**Comment by dscho** [comment]
Sadly, [the AppVeyor build still fails](https://ci.appveyor.com/project/curlorg/curl/builds/31259860), but it is due to a timeout (the `autotools` build runs right into the 1h cut-off, although it reports that the new test 2070 is skipped due to the missing Secure Channel support).

So I think this PR is good to go?

**Comment by mback2k** [comment]
@dscho correct me if I am wrong, but doesn't our schannel implementation nowadays support custom CA bundles? So why do we need to skip the test on MinGW?
https://github.com/curl/curl/blob/32a28dc28d54c73d9db2424763c45da7696dc0ac/lib/vtls/schannel.c#L463-L477
Or is the classic MinGW environment to old? Maybe the test works with mingw-w64 environments?

File: tests/runtests.pl [hunk 2]
```diff
@@ -245,6 +245,7 @@ BEGIN
 my $has_multissl;   # set if curl is build with MultiSSL support
 my $has_manual;     # set if curl is built with built-in manual
 my $has_win32;      # set if curl is built for Windows
+my $has_mingw;      # set if curl is built with MinGW (as opposed to MinGW-w64)
```
Comment (mback2k) [inline tests/runtests.pl]: maybe we can use the chance to get a feature for classic MinGW and one for mingw-w64?

**Comment by dscho** [comment]
@mback2k my guess is as good as yours; I do not have an MSys1 environment to test anymore. But from my understanding [that error message is generated only in the MinGW build](https://github.com/curl/curl/blob/52d302ed64822347de132e50027aaa35f23b145b/lib/vtls/schannel.c#L477-L482) (_not_ the `mingw-w64` one), as indicated by `HAS_MANUAL_VERIFY_API` not being defined.

At least I see [that constant](https://github.com/curl/curl/blob/832661b3a77cef6ba069f6414489c3e94d5d489c/lib/vtls/schannel.h#L64) being guarded by `#ifdef __MINGW64_VERSION_MAJOR` (which means: "is this mingw-w64"?). My guess is that the Win32 API headers shipped with MinGW are simply too old to include the necessary stuff.

File: tests/runtests.pl (hunk 2 above)
Comment (dscho) [reply tests/runtests.pl]: I had that locally, but decided against including it if I don't need it...

**Comment by mback2k** [comment]
@dscho ok, that seems totally plausible. Let's wait for the travis-ci and AppVeyor runs to finish.

File: tests/runtests.pl (hunk 2 above)
Comment (mback2k) [reply tests/runtests.pl]: ok, maybe you can do a 2nd PR for this later on? if you had it working already, I think it would be good to have it for future tests.

Comment (dscho) [reply tests/runtests.pl]: Isn't it too easy as an add-on when needed? I mean, it really is just copy-editing all the `has_mingw` parts, replacing the `-pc-mingw32` needle with a `-w64-mingw32` one (yes, also for 64-bit the suffix is `-w64-mingw32`).

When I read your suggestion, I also considered reversing the logic so that `mingw-w64` would be required, rather than `!MinGW`. But then I realized that the test runs just fine in the Visual Studio builds, so that would be too restrictive.

**Comment by dscho** [comment]
> Let's wait for the travis-ci and AppVeyor runs to finish.

Took long enough for Travis to even _queue_ it. But now [it failed](https://travis-ci.org/curl/curl/builds/658608249?utm_source=github_status&utm_medium=notification), in three jobs no less.

As far as I can see, they are all bogus:
- [13262.18](https://travis-ci.org/curl/curl/jobs/658608269?utm_medium=notification&utm_source=github_status) fails in `./configure` with "Check your system clock"!
- [13262.25](https://travis-ci.org/curl/curl/jobs/658608276?utm_medium=notification&utm_source=github_status) fails before `apt-get` can even retrieve packages because it could not retrieve a GPG key.
- [13262.31](https://travis-ci.org/curl/curl/jobs/658608282?utm_medium=notification&utm_source=github_status) apparently sees random processes being Killed during the test run.

@mback2k could I ask you to re-run those bogusly-failed jobs?

**Comment by dscho** [comment]
@mback2k thank you! Now they all pass! 🚀 🎉

**Comment by dscho** [comment]
@jay anything else you want me to address?

**Comment by jay** [comment]
> anything else you want me to address?

No it looks done. It's tagged for the next feature window.

**Comment by dscho** [comment]
Thank you all!

**Comment by jay** [comment]
Thanks

**Comment by dscho** [comment]
Thank you, all!

**Comment by thniig** [comment]
Network is locked so i cant change anything on it plus dms are also flakey. How to preceed?

**Comment by bagder** [comment]
@thniig you're commenting on a PR that's been closed for months. It won't help you even if we would understand what you're talking about, which I don't. If you have a (lib)curl bug, file an issue. If you have a question, post it on the suitable curl mailing list.
